graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java

```diff
@@ -0,0 +1,207 @@
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.entitygroups.rest;
+
+import com.mongodb.DuplicateKeyException;
+import com.mongodb.MongoWriteException;
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import jakarta.ws.rs.BadRequestException;
+import jakarta.ws.rs.ForbiddenException;
+import jakarta.ws.rs.NotFoundException;
+import org.apache.shiro.authz.annotation.RequiresAuthentication;
+import org.apache.shiro.authz.annotation.RequiresPermissions;
+
+import org.graylog2.audit.AuditEventTypes;
+import org.graylog2.audit.jersey.AuditEvent;
+import org.graylog2.audit.jersey.NoAuditEvent;
+import org.graylog2.entitygroups.EntityGroupService;
+import org.graylog2.entitygroups.model.EntityGroup;
+import org.graylog2.database.PaginatedList;
+import org.graylog2.plugin.rest.PluginRestResource;
+import org.graylog2.rest.models.SortOrder;
+import org.graylog2.shared.rest.resources.RestResource;
+
+import jakarta.inject.Inject;
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.DELETE;
+import jakarta.ws.rs.DefaultValue;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.PUT;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.PathParam;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+import org.graylog2.shared.utilities.StringUtils;
+
+import static org.graylog2.entitygroups.rest.EntityGroupPermissions.ENTITY_GROUP_CREATE;
+import static org.graylog2.entitygroups.rest.EntityGroupPermissions.ENTITY_GROUP_DELETE;
+import static org.graylog2.entitygroups.rest.EntityGroupPermissions.ENTITY_GROUP_EDIT;
+import static org.graylog2.entitygroups.rest.EntityGroupPermissions.ENTITY_GROUP_READ;
+import static org.graylog2.shared.utilities.ExceptionUtils.getRootCauseMessage;
+
+@Api(value = "EntityGroups", description = "Manage Entity Groups")
+@Path("/entity_groups")
+@RequiresAuthentication
+@Produces(MediaType.APPLICATION_JSON)
+public class EntityGroupResource extends RestResource implements PluginRestResource {
+
+    private final EntityGroupService entityGroupService;
+
+    @Inject
+    public EntityGroupResource(EntityGroupService entityGroupService) {
+        this.entityGroupService = entityGroupService;
+    }
+
+    @GET
+    @ApiOperation(value = "Get a list of entity groups")
+    @RequiresPermissions(ENTITY_GROUP_READ)
```

Comment:
In resources that return a list of entities to a user, the entities need to be permission-checked before being returned to the client.

While this creates an awkward problem with pagination and limits, it's currently the only way to ensure we don't leak information.

One way to do this here would be:

```java
final PaginatedList<EntityGroup> paginated = entityGroupService.findPaginated(query, page, perPage, order, sort, null);
return paginated.withList(paginated.stream().filter(entityGroup -> isPermitted(ENTITY_GROUP_READ, entityGroup.id())).toList());
```

This approach isn't great, as it could have completely empty pages or pages with less than `per_page` entries, but at least it would be correct.
There's code in the search classes that extends pagination with a predicted, but for now there's no globally available method to do this properly.

Reply:
In fact there's https://github.com/Graylog2/graylog2-server/blob/1967c43ac4e4e2c9a4ab37517c9cb7cb4fffe3fb/graylog2-server/src/main/java/org/graylog2/database/pagination/MongoPaginationHelper.java#L104 which can help do this already

Reply:
Definitely makes sense, went ahead and refactored the paginated endpoints to use the MongoPaginationHelper filters downstream.